consensus/XDPoS/engines/engine_v2/engine.go

```diff
@@ -40,3 +42,196 @@ func (consensus *XDPoS_v2) Author(header *types.Header) (common.Address, error)
+	*/
+}
+
+/*
+	Timeout workflow
+*/
+// Verify timeout message type from peers in bft.go
+func (consensus *XDPoS_v2) VerifyTimeoutMessage() error {
+	/*
+		  1. Check signature:
+					- Use ecRecover to get the public key
+					- Use the above public key to find out the xdc address
+					- Use the above xdc address to check against the master node(For the running epoch)
+			2. Broadcast(Not part of consensus)
+	*/
+	return nil
+}
+
+func (consensus *XDPoS_v2) TimeoutHandler() {
+	/*
+		1. checkRoundNumber()
+		2. Collect timeout (TODO)
+		3. Genrate TC (TODO)
+		4. processTC()
+		5. generateSyncInfo()
+	*/
+}
+
+/*
+	Process Block workflow
+*/
+func (consensus *XDPoS_v2) ProcessBlockHandler() {
+	/*
+		1. processQC()
+		2. verifyVotingRule()
+		3. sendVote()
+
+	*/
+}
+
+/*
+	QC & TC Utils
+*/
+
+// Genrate blockInfo which contains Hash, round and blockNumber and send to queue
+func (consensus *XDPoS_v2) generateBlockInfo() error {
+	return nil
+}
+
+// To be used by different message verification. Verify local DB block info against the received block information(i.e hash, blockNum, round)
+func (consensus *XDPoS_v2) veifyBlockInfo(header *types.Header) error {
+	return nil
+}
+
+func (consensus *XDPoS_v2) veifyQC(header *types.Header) error {
+	/*
+		1. Verify signer signatures: (List of signatures)
+					- Use ecRecover to get the public key
+					- Use the above public key to find out the xdc address
+					- Use the above xdc address to check against the master node list(For the received QC epoch)
+		2. Verify blockInfo
+	*/
+	return nil
+}
+
+func (consensus *XDPoS_v2) veifyTC(header *types.Header) error {
+	/*
+		1. Verify signer signature: (List of signatures)
+					- Use ecRecover to get the public key
+					- Use the above public key to find out the xdc address
+					- Use the above xdc address to check against the master node list(For the received TC epoch)
+	*/
+	return nil
+}
+
+// Update local QC variables including highestQC & lockQC, as well as update commit blockInfo before call
+func (consensus *XDPoS_v2) processQC(header *types.Header) error {
+	/*
+		1. Update HighestQC and LockQC
+		2. Update commit block info (TODO)
+		3. Check QC round >= node's currentRound. If yes, call setNewRound
+	*/
+	return nil
+}
+
+func (consensus *XDPoS_v2) processTC(header *types.Header) error {
+	/*
+		1. Update highestTC
+		2. Check TC round >= node's currentRound. If yes, call setNewRound
+	*/
+	return nil
+}
+
+func (consensus *XDPoS_v2) setNewRound() error {
+	/*
+		1. Set currentRound = QC round + 1 (or TC round +1)
+		2. Reset timer
+		3. Reset vote and timeout Pools
+	*/
+	return nil
+}
+
+// Verify round number against node's local round number(Should be equal)
+func (consensus *XDPoS_v2) checkRoundNumber(header *types.Header) error {
+	return nil
+}
+
+// Hot stuff rule to decide whether this node is eligible to vote for the received block
+func (consensus *XDPoS_v2) verifyVotingRule(header *types.Header) error {
+	/*
+		(TODO)
```

Comment:
I can write this

Reply:
1. Make sure this node has not voted for this round. We can have a variable `highestVotedRound`, and check `currentRound > highestVotedRound`.
2. HotStuff Voting rule:
- header's round == local current round, AND (one of the following two:)
- header's block extends LockQC's ProposedBlockInfo OR (we need a isExtending(block_a, block_b) function)
- header's QC's ProposedBlockInfo.Round > LockQC's ProposedBlockInfo.Round